src/protocol.rs

```diff
@@ -0,0 +1,760 @@
+            ReadReg => {
+                let address = u32_from_slice(payload, CMD_BASE_SIZE);
+                response.value(read_register(address));
+            }
+            WriteReg => {
+                let reg: WriteRegCommand = slice_to_struct(payload)?;
+                write_register(reg.address, reg.value);
+            }
+            FlashBegin | MemBegin | FlashDeflBegin => {
+                let cmd: BeginCommand = slice_to_struct(payload)?;
+                self.process_begin(&cmd)?
+            }
+            FlashData | FlashDeflData | FlashEncryptedData | MemData => {
+                let cmd: DataCommand = slice_to_struct(&payload)?;
+                let data = &payload[DATA_CMD_SIZE..];
+                self.process_data(&cmd, data, &response)?;
+                response_sent = true;
+            }
+            FlashEnd | MemEnd | FlashDeflEnd => {
+                let cmd: EndCommand = slice_to_struct(payload)?;
+                self.process_end(&cmd, &response)?;
+            }
+            SpiFlashMd5 => {
+                let cmd: SpiFlashMd5Command = slice_to_struct(payload)?;
+                let md5 = calculate_md5(cmd.address, cmd.size)?;
+                self.send_md5_response(&response, &md5);
+                response_sent = true;
+            }
+            SpiSetParams => {
+                let cmd: SpiSetParamsCommand = slice_to_struct(payload)?;
+                spi_set_params(&cmd.params)?
+            }
+            SpiAttach => {
+                let param = u32_from_slice(payload, CMD_BASE_SIZE);
+                spi_attach(param);
+            }
+            ChangeBaudrate => {
+                let baud: ChangeBaudrateCommand = slice_to_struct(payload)?;
+                self.send_response(&response);
+                delay_us(10000); // Wait for response to be transfered 
+                change_baudrate(baud.old, baud.new);
+                self.send_greeting();
+                response_sent = true;
+            }
+            EraseFlash => {
+                erase_flash()?
+            }
+            EraseRegion => {
+                let reg: EraseRegionCommand = slice_to_struct(payload)?;
+                erase_region(reg.address, reg.size)?;
+            }
+            ReadFlash => {
+                self.send_response(&response);
+                let cmd: ReadFlashCommand = slice_to_struct(payload)?;
+                self.process_read_flash(&cmd.params)?;
+                response_sent = true;
+            }
+            GetSecurityInfo => {
+                let info = get_security_info()?;
+                self.send_security_info_response(&response, &info);
+                response_sent = true;
+            }
+            RunUserCode => {
+                soft_reset(); // ESP8266 Only
+            }
+            _ => {
+                return Err(InvalidCommand);
+            }
+        };
+
+        Ok(response_sent)
+    }
+
+    pub fn process_command(&mut self, payload: &[u8]) {
```

Comment:
I'm not a huge fan of having ```process_cmd``` and ```process_command``` function names, they are too similar and the difference between them cannot be discerned from the name.

Reply:
Me neither :blush:. I just couldn't come up with something more descriptive. `process_cmd` -> `execute_command`